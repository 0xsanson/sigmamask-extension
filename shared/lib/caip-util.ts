export function isCaipChainId(caipChainId: string): boolean {
  return caipChainId.startsWith('eip155:');
}
export function getCaipChainIdFromEthChainId(ethChainId: string): string {
  const chainIdDecimal = ethChainId.startsWith('0x')
    ? parseInt(ethChainId, 16).toString(10)
    : ethChainId;
  return `eip155:${chainIdDecimal}`;
}

export function getEthChainIdDecFromCaipChainId(caipChainId: string): string {
  return caipChainId.replace(/^eip155:/, '')
}

export function getEthChainIdHexFromCaipChainId(caipChainId: string): string {
  const chainIdDecimal = getEthChainIdDecFromCaipChainId(caipChainId)
  return `0x${parseInt(chainIdDecimal, 10).toString(16)}`;
}

export function getEthChainIdNumFromCaipChainId(caipChainId: string): number {
  const chainIdDecimal = getEthChainIdDecFromCaipChainId(caipChainId)
  return parseInt(chainIdDecimal, 10)
}
