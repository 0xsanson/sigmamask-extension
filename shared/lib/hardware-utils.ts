import { isManifestV3 } from '../modules/mv3.utils';
import { LedgerTransportTypes } from '../constants/hardware-wallets';

export const getDefaultLedgerTransportType = (): LedgerTransportTypes => {
  if (isManifestV3) {
    // Then enforce the use of WebHID as the default, despite
    // the current context not supporting window.navigator.hid
    return LedgerTransportTypes.webhid;
  }

  return Object.hasOwnProperty.call(window.navigator, 'hid')
    ? LedgerTransportTypes.webhid
    : LedgerTransportTypes.u2f;
};

export const getRecommendedLedgerTransportType = (): LedgerTransportTypes => {
  // Currently the recommended ledger transport type is the default
  return getDefaultLedgerTransportType();
};

type LedgerTranslations = Record<LedgerTransportTypes, string>;
type LedgerTransportOption = {
  name: LedgerTranslations[LedgerTransportTypes];
  value: LedgerTransportTypes;
};

export const getLedgerTransportOption = (
  transportType: LedgerTransportTypes,
  translations: LedgerTranslations,
): LedgerTransportOption => {
  return {
    name: translations[transportType],
    value: transportType,
  };
};
