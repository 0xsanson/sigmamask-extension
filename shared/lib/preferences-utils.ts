import { isManifestV3 } from '../modules/mv3.utils';
import { LedgerTransportTypes } from '../constants/hardware-wallets';

export const getInitLedgerTransportType = (): string => {
  if (isManifestV3) {
    // Then enforce the use of WebHID as the default, despite
    // the current context not supporting window.navigator.hid
    return LedgerTransportTypes.webhid;
  }

  return Object.hasOwnProperty.call(window.navigator, 'hid')
    ? LedgerTransportTypes.webhid
    : LedgerTransportTypes.u2f;
};
