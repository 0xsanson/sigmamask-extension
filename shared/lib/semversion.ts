/**
 * A simple utility to parse and compare semver versions.
 */

import type ExtensionPlatform from '../../app/scripts/platforms/extension';

/**
 * A semver version object.
 */
export type SemVersion = {
  major: number;
  minor: number;
  patch: number;
};

/**
 * Parse a semver version string into a SemVersion object.
 *
 * @param version - The version string to parse, of the form "major.minor.patch". e.g. "1.2.3"
 * @returns
 */
export const parseVersion = (version: string = ''): SemVersion | null => {
  const [major, minor, patch] = version.split('.').map(Number);

  // if version is not a valid semver, return false
  if ([major, minor, patch].some((num) => isNaN(num) || num < 0)) {
    return null;
  }
  return { major, minor, patch };
};

/**
 * Compare two SemVersion objects.
 *
 * @param a
 * @param b
 * @returns negative if a < b, zero if a === b, positive if a > b
 */
export const compareVersions = (a: SemVersion, b: SemVersion): number => {
  if (a.major !== b.major) {
    return a.major - b.major;
  }
  if (a.minor !== b.minor) {
    return a.minor - b.minor;
  }
  return a.patch - b.patch;
};

/**
 * Get the version of the extension.
 */
export const getMetamaskVersion = () =>
  parseVersion((global.platform as ExtensionPlatform).getVersion());
