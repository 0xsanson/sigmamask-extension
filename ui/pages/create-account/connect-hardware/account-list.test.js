import React from 'react';
import configureMockStore from 'redux-mock-store';
import {
  HardwareDeviceNames,
  DEFAULT_HD_PATHS,
  HD_PATHS,
} from '../../../../shared/constants/hardware-wallets';
import { renderWithProvider } from '../../../../test/lib/render-helpers';
import AccountList from './account-list';

const accountId = '0x8eeee1781fd885ff5ddef7789486676961873d12';
const mockState = {
  metamask: {
    provider: { chainId: '0x5', type: 'goerli' },
    keyrings: [],
    cachedBalances: {
      '0x5': {},
    },
    accounts: {
      [accountId]: {
        address: accountId,
        balance: '0x7e64033f2fdb0436',
        index: 0,
      },
    },
    selectedAddress: accountId,
  },
  appState: {
    defaultHdPaths: DEFAULT_HD_PATHS,
  },
};

describe('Account List', () => {
  const store = configureMockStore()(mockState);

  describe('render', () => {
    it('should match snapshot', () => {
      const { [accountId]: selectedAccount } = mockState.metamask.accounts;
      // eslint-disable-next-line no-empty-function
      const noOp = () => {};
      const render = () => (
        <AccountList
          device={HardwareDeviceNames.ledger}
          hdPaths={HD_PATHS}
          accounts={[selectedAccount]}
          selectedAccounts={[selectedAccount]}
          getPage={noOp}
          onAccountChange={noOp}
          onAccountRestriction={noOp}
          onCancel={noOp}
          onForgetDevice={noOp}
          onPathChange={noOp}
          onUnlockAccounts={noOp}
        ></AccountList>
      );
      const { container } = renderWithProvider(render(), store);
      expect(container).toMatchSnapshot();
    });
  });
});
