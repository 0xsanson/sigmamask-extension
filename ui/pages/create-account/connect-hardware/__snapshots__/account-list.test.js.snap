// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Account List render should match snapshot 1`] = `
<div>
  <div
    class="new-external-account-form account-list"
  >
    <div
      class="hw-connect"
    >
      <h3
        class="hw-connect__unlock-title"
      >
        Select an account
      </h3>
      <div>
        <h3
          class="hw-connect__hdPath__title"
        >
          Select HD path
        </h3>
        <p
          class="hw-connect__msg"
        >
          If you don't see the accounts you expect, try switching the HD path.
        </p>
        <div
          class="hw-connect__hdPath"
        >
          <div
            class="dropdown hw-connect__hdPath__select"
          >
            <select
              class="dropdown__select"
            >
              <option
                value="m/44'/60'/0'/0/0"
              >
                Ledger Live
              </option>
              <option
                value="m/44'/60'/0'"
              >
                Legacy (MEW / MyCrypto)
              </option>
              <option
                value="m/44'/60'/0'/0"
              >
                BIP44 Standard (e.g. MetaMask, Trezor)
              </option>
            </select>
            <span
              class="box dropdown__icon-caret-down mm-icon mm-icon--size-sm box--display-inline-block box--flex-direction-row box--color-inherit"
              style="mask-image: url('./images/icons/arrow-down.svg');"
            />
          </div>
        </div>
      </div>
      <h3
        class="hw-connect__hdPath__title"
      >
        Select an account
      </h3>
    </div>
    <div
      class="hw-account-list"
    >
      <div
        class="hw-account-list__item"
        title=""
      >
        <div
          class="hw-account-list__item__checkbox"
        >
          <input
            class="check-box far fa-square"
            id="address-0"
            readonly=""
            type="checkbox"
          />
          <label
            class="hw-account-list__item__label"
            for="address-0"
          >
            <span
              class="hw-account-list__item__index"
            >
              1
            </span>
            0x8e...3d12
            <span
              class="hw-account-list__item__balance"
            >
              9.107407 ETH
            </span>
          </label>
        </div>
        <a
          class="hw-account-list__item__link"
          rel="noopener noreferrer"
          target="_blank"
          title="View account on Etherscan"
        >
          <i
            class="fa fa-share-square"
          />
        </a>
      </div>
    </div>
    <div
      class="hw-list-pagination"
    >
      <button
        class="hw-list-pagination__button"
      >
        &lt; Prev
      </button>
      <button
        class="hw-list-pagination__button"
      >
        Next &gt;
      </button>
    </div>
    <div
      class="new-external-account-form__buttons"
    >
      <button
        class="button btn--rounded btn-secondary btn--large new-external-account-form__button"
        role="button"
        tabindex="0"
      >
        Cancel
      </button>
      <button
        class="button btn--rounded btn-primary btn--large new-external-account-form__button unlock"
        role="button"
        tabindex="0"
      >
        Unlock
      </button>
    </div>
    <div
      class="hw-forget-device-container"
    >
      <a>
        Forget this device
      </a>
    </div>
  </div>
</div>
`;
